### PR TITLE
Generic MySQL credentials provider

### PR DESCRIPTION
**Pull request by byewokko**
Related to #35 

## Read-only provider

- MySQL provider is read-only by default
- `list`, `get` and `locate` queries must be specified
- `AS` clauses must be used to map the result fields to expected Seacat Auth fields `_id`, `username`, `email`, `phone` and `__password`

Example config:
```ini
[seacatauth:credentials:mysql:external]
user=root
password=rootpassword
database=auth
data_fields=firstName lastName
list=
    SELECT `id` AS '_id', `userName` AS 'username', `userEmail` AS 'email', `userPhone` AS 'phone'
    FROM `users` 
    ORDER BY `id` ASC;
get=
    SELECT `id` AS '_id', `userName` AS 'username', `userEmail` AS 'email', `userPhone` AS 'phone', `userPwd` AS '__password',  `userSuspened` AS 'suspended'
    FROM `users` 
    WHERE `id` = %(_id)s;
locate=
    SELECT `id` AS '_id', `userName`, `userEmail`
    FROM `users` 
    WHERE (LOWER(`userName`) = %(ident)s) OR (LOWER(`userEmail`) = %(ident)s);
```

## Editable provider

- To enable creating, updating and deleting users, specify `editable=yes` in config
- This requires specifying the `create`, `update` and `delete` queries

Example config:
```ini
[seacatauth:credentials:mysql:external]
editable=yes
user=root
password=rootpassword
database=auth
data_fields=firstName lastName
list=
    SELECT `id` AS '_id', `userName` AS 'username', `userEmail` AS 'email', `userPhone` AS 'phone'
    FROM `users` 
    ORDER BY `id` ASC;
get=
    SELECT `id` AS '_id', `userName` AS 'username', `userEmail` AS 'email', `userPhone` AS 'phone', `userPwd` AS '__password',  `userSuspened` AS 'suspended'
    FROM `users` 
    WHERE `id` = %(_id)s;
locate=
    SELECT `id` AS '_id', `userName`, `userEmail`
    FROM `users` 
    WHERE (LOWER(`userName`) = %(ident)s) OR (LOWER(`userEmail`) = %(ident)s);
create=
    INSERT INTO `users` 
    (`userName`, `userEmail`, `userPhone`) 
    VALUES (%(username)s, %(email)s, %(phone)s);
update=
    UPDATE `users` 
    SET `userEmail` = %(email)s, `userPhone` = %(phone)s, `userSuspened` = %(suspended)s, `userPwd` = %(__password)s
    WHERE `id` = %(_id)s;
delete=
    DELETE FROM `users` 
    WHERE `id` = %(_id)s;
```

### CODE REVIEW

File: seacatauth/credentials/providers/mysql.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+		self.PasswordField = self.Config.get("field_password")
+
+		data_fields = self.Config.get("data_fields")
+		if len(data_fields) > 0:
+			self.DataFields = data_fields.split(" ")
+		else:
+			self.DataFields = None
+
+
+	async def create(self, credentials: dict) -> Optional[str]:
```
Comment (ateska) [inline seacatauth/credentials/providers/mysql.py:81]: Start with the read-only provider first.

File: seacatauth/credentials/providers/mysql.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+			async with connection.cursor(aiomysql.DictCursor) as cursor:
+				await cursor.execute(query)
+				result = await cursor.fetchone()
+		if result is None:
+			return None
+		return "{}{}".format(self.Prefix, result[self.IdField])
+
+
+	async def get_by(self, key: str, value) -> Optional[dict]:
+		query = "SELECT * FROM `{table}` WHERE `{field}` = {value};".format(
```
Comment (ateska) [inline seacatauth/credentials/providers/mysql.py:199]: `SELECT *` is a problematic construct.
You should let people to specify fields to select in the configuration - take advantage of "... AS ..." if needed.

**Review by ateska — CHANGES_REQUESTED** [review]

File: seacatauth/credentials/providers/mysql.py (hunk 1 above)
Comment (byewokko) [reply seacatauth/credentials/providers/mysql.py:83]: Provider is read-only by default. Set `editable=yes` to enable editing.

**Review by ateska — APPROVED** [review]
Looks good.
You need to provide a samples of queries - in the documentation.

**Pull request merged**